Sources/Regex/Parse/LexicalAnalysis.swift

```diff
@@ -0,0 +1,456 @@
+      // But Java 8 errors out?
+      if src.tryEat("{") {
+        // FIXME: Erm, PCRE parses as literal if no lowerbound...
+        let amt = try src.expectRange()
+        let kind = consumeKind()
+        try src.expect("}")
+        return Quantifier(amt.value, kind, nil)
+      }
+
+      return nil
+    }
+  }
+
+  /// Consume a range
+  ///
+  ///     Range       -> ',' <Int> | <Int> ',' <Int>? | <Int>
+  ///                  | ModernRange
+  ///     ModernRange -> '..<' <Int> | '...' <Int>
+  ///                  | <Int> '..<' <Int> | <Int> '...' <Int>?
+  mutating func expectRange() throws -> Value<Quantifier.Amount> {
+    try recordLoc { src in
+      // TODO: lex positive numbers, more specifically...
+
+      let lowerOpt = try src.lexNumber()?.value
+
+      // ',' or '...' or '..<' or nothing
+      let closedRange: Bool?
+      if src.tryEat(",") {
+        closedRange = true
+      } else if src.modernRanges && src.tryEat(".") {
+        try src.expect(".")
+        if src.tryEat(".") {
+          closedRange = true
+        } else {
+          try src.expect("<")
+          closedRange = false
+        }
+      } else {
+        closedRange = nil
+      }
+      let upperOpt = try! src.lexNumber()?.value
+
+      switch (lowerOpt, closedRange, upperOpt) {
+      case let (l?, nil, nil):
+        return .exactly(l)
+      case let (l?, true, nil):
+        return .nOrMore(l)
+      case let (nil, closed?, u?):
+        return .upToN(closed ? u : u-1)
+      case let (l?, closed?, u?):
+        return .range(l...(closed ? u : u-1))
+      case let (nil, nil, u) where u != nil:
+        fatalError("Not possible")
+      default:
+        throw LexicalError.misc("Invalid range")
+      }
+    }
+  }
+
+  /// Expect a linear run of non-nested non-empty content
+  private mutating func expectQuoted(
+    endingWith end: String
+  ) throws -> Value<String> {
+    try recordLoc { src in 
+      var result = ""
+      while !src.tryEat(sequence: end) {
+        // TODO(diagnostic): expected `end`, instead of end-of-input
+
+        result.append(src.eat())
+      }
+
+      guard !result.isEmpty else {
+        throw LexicalError.misc("Expected non-empty contents")
+      }
+      return result
+    }
+  }
+
+  /// Try to consume quoted content
+  ///
+  ///     Quote -> '\Q' (!'\E' .)* '\E'
+  ///
+  /// With `SyntaxOptions.modernQuotes`, also accepts
+  ///
+  ///     ModernQuote -> '"' [^"]* '"'
+  ///
+  /// Future: Modern quotes are full fledged Swift string literals
+  ///
+  /// TODO: Need to support some escapes
+  ///
+  mutating func lexQuote() throws -> Value<String>? {
+    try recordLoc { src in 
+      if src.tryEat(sequence: #"\Q"#) {
+        return try src.expectQuoted(endingWith: #"\E"#).value
+      }
+      if src.modernQuotes, src.tryEat("\"") {
+        // TODO: escaped `"`, etc...
+        return try src.expectQuoted(endingWith: "\"").value
+      }
```

Comment:
Out of curiosity, would we allow modern quotes in custom character classes? `\Q...\E` are currently allowed and just escape things so `[\Qabc\E]` is equivalent to `[abc]`, but it may be confusing to allow `["abc"]` to be interpreted as `[abc]`.

Reply:
Hmm, another interpretation could be that they spell a single Character, like `"a\u{301}"`.

Outside of a character class, adjacency means concatenation and inside it's listing set members, so PCRE's quote behavior does make sense. When we add in grapheme-semantics, these things get a lot murkier.

(noting that the "modern syntax" stuff is just experimental) I wouldn't be surprised if we try to restrict it to Character literals, and since we're in Swift and can do grapheme breaking, we could finally issue a static error if it's not a single grapheme cluster.